### PR TITLE
fix(pg-topo): support alter publication subscription

### LINKED ISSUE #283: pg-topo: ALTER PUBLICATION and ALTER SUBSCRIPTION are treated as unsupported
## Summary

`pg-topo` documents publication/subscription support, but `ALTER PUBLICATION` and `ALTER SUBSCRIPTION` statements are currently classified as `UNKNOWN`. Because unknown statements keep only phase/class tie-breaker ordering and no object refs, they can be ordered before the objects they alter.

This is a separate issue class from supabase/pg-toolbelt#281 and supabase/pg-toolbelt#282:

* supabase/pg-toolbelt#281 is about expression dependencies inside `ALTER TABLE` subcommands.
* supabase/pg-toolbelt#282 is about `CREATE TYPE ... AS RANGE` / `CreateRangeStmt` coverage.
* This issue is about missing publication/subscription ALTER statement classification and dependency extraction.

Current `main`: `9f01826b00b64e7f9628844ecb6f69a07f53c1f8`

## Repro A: ALTER PUBLICATION ADD TABLE

Input statements, intentionally shuffled:

```sql
alter publication pub_items add table app.items;

create publication pub_items;

create table app.items(id int primary key);

create schema app;
```

Actual ordered output from the local probe:

```json
{
  "ordered": [
    {
      "class": "CREATE_SCHEMA",
      "sql": "create schema app;"
    },
    {
      "class": "CREATE_TABLE",
      "sql": "create table app.items(id int primary key);"
    },
    {
      "class": "UNKNOWN",
      "sql": "alter publication pub_items add table app.items;"
    },
    {
      "class": "CREATE_PUBLICATION",
      "sql": "create publication pub_items;"
    }
  ],
  "diagnostics": [
    {
      "code": "UNKNOWN_STATEMENT_CLASS",
      "message": "Unsupported statement AST root 'AlterPublicationStmt'."
    }
  ]
}
```

Runtime validation also shows the bad ordering symptom:

```text
publication "pub_items" does not exist
sqlstate: 42704
statementClass: UNKNOWN
```

Expected order:

```text
CREATE_SCHEMA
CREATE_TABLE
CREATE_PUBLICATION
ALTER_PUBLICATION -- add app.items to pub_items
```

## Repro B: ALTER SUBSCRIPTION SET PUBLICATION

Input statements, intentionally shuffled:

```sql
alter subscription sub_items
  set publication pub_a, pub_b
  with (refresh = false);

create subscription sub_items
  connection 'host=localhost port=5432 dbname=postgres'
  publication pub_a
  with (enabled = false, create_slot = false, connect = false);

create publication pub_a;
create publication pub_b;
```

Actual ordered output from the local probe:

```json
{
  "ordered": [
    {
      "class": "UNKNOWN",
      "sql": "alter subscription sub_items set publication pub_a, pub_b with (refresh = false);"
    },
    {
      "class": "CREATE_PUBLICATION",
      "sql": "create publication pub_b;"
    },
    {
      "class": "CREATE_PUBLICATION",
      "sql": "create publication pub_a;"
    },
    {
      "class": "CREATE_SUBSCRIPTION",
      "sql": "create subscription sub_items connection 'host=localhost port=5432 dbname=postgres' publication pub_a with (enabled = false, create_slot = false, connect = false);"
    }
  ],
  "diagnostics": [
    {
      "code": "UNKNOWN_STATEMENT_CLASS",
      "message": "Unsupported statement AST root 'AlterSubscriptionStmt'."
    }
  ]
}
```

The local runtime validator reports `subscription \"sub_items\" does not exist` at the ALTER statement. Subscription validation can have local environment limitations, but the static problem is independent of that: the statement is classified as `UNKNOWN` and has no dependency on either the existing subscription or the newly referenced publications.

Expected order:

```text
CREATE_PUBLICATION pub_a
CREATE_PUBLICATION pub_b
CREATE_SUBSCRIPTION sub_items
ALTER_SUBSCRIPTION -- set publication pub_a, pub_b
```

## Root cause

`packages/pg-topo/src/classify/classify-statement.ts` maps only `CreatePublicationStmt` and `CreateSubscriptionStmt` for these object families. It does not map:

* `AlterPublicationStmt`
* `AlterSubscriptionStmt`

As a result, `extractDependenciesForStatement` never gets a chance to emit object refs such as:

* `ALTER PUBLICATION ... ADD TABLE ...` requires the existing publication and the added table(s).
* `ALTER PUBLICATION ... DROP TABLE ...` requires the existing publication; table refs may still be useful for deterministic diagnostics/order when present.
* `ALTER SUBSCRIPTION ... SET PUBLICATION ...` requires the existing subscription and the referenced publication(s).
* `ALTER SUBSCRIPTION ... ENABLE/DISABLE/CONNECTION/SET (...)` requires the existing subscription.

## Suggested fix direction

1. Add statement classes for publication/subscription ALTER variants, or reuse existing object-family classes if the current model prefers fewer class names.
2. Add extractor coverage for `AlterPublicationStmt`:
   * provide no new object;
   * require the target `publication:*`;
   * require table refs for `ADD TABLE` / `SET TABLE` variants.
3. Add extractor coverage for `AlterSubscriptionStmt`:
   * provide no new object;
   * require the target `subscription:*`;
   * require referenced `publication:*` entries for `SET PUBLICATION` / add publication variants.
4. Add runtime-backed pg-topo tests for `ALTER PUBLICATION ADD TABLE`; for subscription, at minimum add static classifier/extractor tests, and runtime validation where the local test container can support it without false environment failures.

## Local validation

Both focused probes were run twice and produced the same unsupported AST root diagnostics.

```bash
./node_modules/.bin/bun run --cwd packages/pg-topo scripts/run-tests.ts \
  --test-name-pattern "ALTER PUBLICATION ADD TABLE" \
  test/deep-probe.test.ts

./node_modules/.bin/bun run --cwd packages/pg-topo scripts/run-tests.ts \
  --test-name-pattern "ALTER SUBSCRIPTION SET PUBLICATION" \
  test/deep-probe.test.ts
```

### PR DESCRIPTION
**Pull request by xmh1011**
## Summary

Closes #283.

This adds pg-topo support for `ALTER PUBLICATION` and `ALTER SUBSCRIPTION` statements:

- classify `AlterPublicationStmt` / `AlterSubscriptionStmt`
- keep both in the post-data phase
- extract dependencies on the target publication/subscription
- extract referenced tables and schemas for publication changes
- extract referenced publications for subscription changes

`ALTER SUBSCRIPTION` remains statically covered because runtime execution depends on connection environment. `ALTER PUBLICATION ADD TABLE` is covered with PostgreSQL runtime validation.

## RED evidence

Before the fix, both statement roots were classified as `UNKNOWN`, so pg-topo had no object refs or dependency edges for them.

## Validation

```bash
bun run --cwd packages/pg-topo scripts/run-tests.ts test/analyze-and-sort.test.ts --test-name-pattern "ALTER PUBLICATION ADD TABLE"
```

Result: `1 pass / 0 fail` with PostgreSQL runtime validation.

```bash
bun run test:pg-topo
```

Result: `149 pass / 0 fail`.

```bash
bun run format-and-lint
bun run check-types
bun run knip
```

Result: passed. `knip` only reported existing configuration hints.

### CODE REVIEW

File: packages/pg-topo/src/extract/extract-dependencies.ts [hunk 1]
```diff
@@ -514,6 +563,32 @@ const extractCreateSubscriptionDependencies = (
...
+): ExtractDependenciesResult => {
+  const requires: ObjectRef[] = [];
+
+  if (typeof statementNode.subname === "string") {
+    requires.push(
+      createObjectRefFromAst("subscription", statementNode.subname),
+    );
+  }
+
+  if (statementNode.kind === "ALTER_SUBSCRIPTION_SET_PUBLICATION") {
```
[the remark this answers is not in the document]
Comment (xmh1011) [reply packages/pg-topo/src/extract/extract-dependencies.ts]: Fixed in e39f822. `ALTER_SUBSCRIPTION_ADD_PUBLICATION` now uses the same publication-list dependency extraction as SET PUBLICATION, so missing added publications are reported and present publications order before the ALTER.

Validation:
`bun run --cwd packages/pg-topo scripts/run-tests.ts test/analyze-and-sort.test.ts test/diagnostics.test.ts` -> 21 pass / 0 fail
`bun run --cwd packages/pg-topo format-and-lint` -> pass

File: packages/pg-topo/src/extract/extract-dependencies.ts [hunk 2]
```diff
@@ -445,6 +445,55 @@ const extractCreatePublicationDependencies = (
...
+  const requires: ObjectRef[] = [];
+
+  if (typeof statementNode.pubname === "string") {
+    requires.push(createObjectRefFromAst("publication", statementNode.pubname));
+  }
+
+  if (
+    statementNode.action === "AP_AddObjects" ||
+    statementNode.action === "AP_SetObjects"
+  ) {
```
[the remark this answers is not in the document]
Comment (xmh1011) [reply packages/pg-topo/src/extract/extract-dependencies.ts:579]: Fixed in c1c00c8. `ALTER PUBLICATION ... DROP TABLE ...` and `DROP TABLES IN SCHEMA ...` now reuse the publication object dependency extractor, so pg-topo records the dropped table/schema requirement, emits `UNRESOLVED_DEPENDENCY` when it is missing, and adds ordering edges when the object is produced in the input.

Validation:
`./node_modules/.bin/bun run --cwd packages/pg-topo scripts/run-tests.ts test/analyze-and-sort.test.ts test/diagnostics.test.ts` -> 25 pass / 0 fail
`./node_modules/.bin/bun run --cwd packages/pg-topo format-and-lint` -> pass
`./node_modules/.bin/bun run --cwd packages/pg-topo check-types` -> pass

File: packages/pg-topo/src/extract/extract-dependencies.ts [hunk 3]
```diff
@@ -504,6 +504,56 @@ const extractCreatePublicationDependencies = (
...
+): void => {
+  const publicationObjects = Array.isArray(objects) ? objects : [];
+  for (const objectNode of publicationObjects) {
+    const publicationObjSpec = asRecord(
+      asRecord(objectNode)?.PublicationObjSpec,
+    );
+    if (!publicationObjSpec) {
+      continue;
+    }
+    if (publicationObjSpec.pubobjtype === "PUBLICATIONOBJ_TABLE") {
```
[the remark this answers is not in the document]
Comment (xmh1011) [reply packages/pg-topo/src/extract/extract-dependencies.ts]: Fixed in 4b0db1c0d44e7085fe8055ad478a4593624bb22b. `addPublicationObjectDependencies` now handles `PUBLICATIONOBJ_CONTINUATION` by extracting carried table refs from `pubtable` and schema refs from `name`, so continuation entries produce the same dependency requirements as leading publication objects. Added a focused regression for continuation table and schema dependencies. Validation: `cd packages/pg-topo && ../../node_modules/.bin/bun run test` -> 167 pass / 0 fail; root `./node_modules/.bin/bun run format-and-lint`, `check-types`, `knip`, and `git diff --check` all exited 0.

File: packages/pg-topo/src/extract/extract-dependencies.ts [hunk 4]
```diff
@@ -504,6 +504,56 @@ const extractCreatePublicationDependencies = (
...
+      const relation = asRecord(
+        asRecord(publicationObjSpec.pubtable)?.relation,
+      );
+      const tableRef = relationFromRangeVarNode(relation, "table");
```
[the remark this answers is not in the document]
Comment (xmh1011) [reply packages/pg-topo/src/extract/extract-dependencies.ts]: Fixed in 4b0db1c0d44e7085fe8055ad478a4593624bb22b. `ALTER PUBLICATION` table objects now pass `pubtable.whereClause` through the existing `addExpressionDependencies` path, which adds row-filter function requirements and unresolved diagnostics when the filter dependency is absent. Added regressions for the row-filter function edge and unresolved `function:public:is_visible:(unknown)` diagnostic. Validation: `cd packages/pg-topo && ../../node_modules/.bin/bun run test` -> 167 pass / 0 fail; root `./node_modules/.bin/bun run format-and-lint`, `check-types`, `knip`, and `git diff --check` all exited 0.

**Review by avallete — COMMENTED** [review]
Thanks for the contribution — this follows the repo conventions well (changeset, Conventional Commit title, RED evidence, tests across the classify/extract/ordering/diagnostics layers plus a runtime-validated case). The overall structure is right: new statement classes, `post_data` phase matching the CREATE counterparts, adjacent tiebreak weights, and extractors that reuse the existing helpers (`relationFromRangeVarNode`, `addExpressionDependencies`, `extractStringValue`). The action/kind gating also looks correct (`AP_SetOptions` only requires the publication itself; `ALTER SUBSCRIPTION ... DROP PUBLICATION` deliberately adds no publication edge, consistent with the `CREATE SUBSCRIPTION` convention).

A few things before this is ready:

**1. The `PUBLICATIONOBJ_CONTINUATION` branches are dead code, and the new unit test exercises ASTs the parser can't produce.**

PostgreSQL's grammar resolves continuations at parse time (`preprocess_pubobj_list` in `gram.y`), so they never reach pg-topo. Verified against this package's actual parser:

- `alter publication p add table a, b;` → two `PUBLICATIONOBJ_TABLE` nodes
- `alter publication p add tables in schema s1, s2;` → two `PUBLICATIONOBJ_TABLES_IN_SCHEMA` nodes

The hand-crafted AST fixtures in `test/alter-publication-dependencies.test.ts` test branches no real input hits. Suggest dropping the continuation handling and that synthetic-AST test, and covering the multi-table/multi-schema cases via `analyzeAndSort` with real SQL like the other new tests do.

**2. Reuse opportunity that closes a real gap in the CREATE path.**

The new `addPublicationObjectDependencies` handles `whereClause` (row-filter expression deps) and `TABLES_IN_SCHEMA`, but the pre-existing `extractCreatePublicationDependencies` still has its own inline loop that handles neither — `create publication p for table b where (x > 1)` does emit a `whereClause` in the AST, so today `CREATE PUBLICATION` with a row filter misses the dependency edge that the new `ALTER PUBLICATION` path catches. Refactoring the create extractor to call the new helper removes the duplication and fixes that gap in one go. Fine as a follow-up issue if you'd rather keep scope tight, but it's a small change and this PR already contains most of the test machinery for it.

**3. Minor test nits.**

- `requires.map(objectRefKey).sort()` followed by `expect.arrayContaining([...])` — the `.sort()` is a no-op with `arrayContaining`. Either drop the sort or assert the exact sorted array (the latter is stronger and catches unexpected extra refs).

**4. Doc drift.**

`packages/pg-topo/CLAUDE.md` says "Statement classification (38 types)" — this PR makes it 40; please update it in the same PR.

**5. Branch is behind `main`** — needs an update before merge, though since this PR is pg-topo-only the recent pg-delta changes on main shouldn't conflict.

Risk is low overall: the change is additive (previously-`UNKNOWN` statements gain edges, nothing existing reorders) and the new `requires` edges all point at objects the statement genuinely references.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by xmh1011** [comment]
Addressed the latest top-level review.

Changes:
- Rebased/synced PR #289 onto latest `origin/main` (`d706336a6772318e92db419eda5a5ea51123510e`).
- Removed dead `PUBLICATIONOBJ_CONTINUATION` handling from publication object dependency extraction and replaced the synthetic AST test with real SQL `analyzeAndSort` coverage for multi-table and multi-schema `ALTER PUBLICATION`.
- Reused `addPublicationObjectDependencies` for `CREATE PUBLICATION`, so row-filter function dependencies and `FOR TABLES IN SCHEMA` schema dependencies are tracked consistently.
- Tightened dependency assertions to exact sorted object-ref arrays.
- Updated `packages/pg-topo/CLAUDE.md` statement classification count from 38 to 40.

RED evidence:
- Before the extractor fix, `DOCKER_HOST=unix:///Users/xiaominghao/.docker/run/docker.sock ../../node_modules/.bin/bun run test --test-name-pattern "CREATE PUBLICATION"` failed because `CREATE PUBLICATION` was missing `function:public:is_visible:(unknown)` and `schema::marketing:` / `schema::sales:` dependency edges.

Validation:
- `DOCKER_HOST=unix:///Users/xiaominghao/.docker/run/docker.sock ./node_modules/.bin/bun run --filter '@supabase/pg-topo' test --test-name-pattern "PUBLICATION|SUBSCRIPTION"` -> 17 pass, 0 fail.
- `cd packages/pg-topo && DOCKER_HOST=unix:///Users/xiaominghao/.docker/run/docker.sock ../../node_modules/.bin/bun run test` -> 175 pass, 0 fail.
- `./node_modules/.bin/bun run format-and-lint` -> pass.
- `./node_modules/.bin/bun run check-types` -> pass.
- `./node_modules/.bin/bun run knip` -> pass (configuration hints only).
- `diff-check` not run: no matching repo script/command found.

Pushed to fork branch at `bffb159e0888c66fad6aa543df8530fbcc78f84c`.

**Review by avallete — APPROVED** [review]
